copy/copy.go

```diff
@@ -394,17 +396,36 @@ func shortDigest(d digest.Digest) string {
 	return d.Encoded()[:12]
 }
 
-// createProgressBar creates a pb.ProgressBar.
-func createProgressBar(srcInfo types.BlobInfo, kind string, writer io.Writer) *pb.ProgressBar {
+// createProgressBar creates a pb.ProgressBar.  Note that if the copier's
+// reportWriter is a file but a tty, the progress bar's output will be
+// discarded and a single line will be printed instead.
+func (c *copier) createProgressBar(srcInfo types.BlobInfo, kind string) *pb.ProgressBar {
 	bar := pb.New(int(srcInfo.Size)).SetUnits(pb.U_BYTES)
 	bar.SetMaxWidth(80)
 	bar.ShowTimeLeft = false
 	bar.ShowPercent = false
-	bar.Prefix(fmt.Sprintf("Copying %s %s:", kind, shortDigest(srcInfo.Digest)))
-	bar.Output = writer
+	prefix := fmt.Sprintf("Copying %s %s:", kind, shortDigest(srcInfo.Digest))
+	bar.Prefix(prefix)
+	if isFileButNoTTY(c.reportWriter) {
+		bar.Output = c.reportWriter
+	} else {
+		bar.Output = ioutil.Discard
+		c.Printf("Copying %s %s\n", kind, srcInfo.Digest)
+	}
 	return bar
 }
 
+// isFileButNoTTY returns true if the io.Writer is a file but not a tty.
+func isFileButNoTTY(w io.Writer) bool {
+	if reflect.TypeOf(w) == reflect.TypeOf((*os.File)(nil)) {
+		f := w.(*os.File)
```

Comment:
Aren’t these two lines basically
```go
if f, ok := w.(*os.File); ok {
```
?

Reply:
Yes, that's much nicer. I keep forgetting that go supports type assertions.